RbShift modifications to use in tests

mijaros opened the pull request.

mijaros committed:
Rbshift now properly uses oc cli

Adding replication controller to check whether deployment
is running of completed.

mijaros committed:
Adds support for inplace reload of object

* The method reload is added to OpenshiftKind
* This is usable for inplace reload of object
  without need to access or invalidate parent element

Signed-off-by: Miroslav Jaros <mirek@mijaros.cz>

jcechace commented on lib/rb_shift/service.rb:
```diff
@@ -7,19 +7,29 @@
-    def create_route(name, termination, **opts)
-      `oc create route #{termination} #{name} --service=#{@name} #{unfold_params(opts)}`
+    def create_route(name, hostname, termination = 'edge', **opts)
+      unless termination.nil?
```
This is a straight up anti pattern and abuse of unless. 

```ruby
if termination
  # do the thing only when termination is NOT nil
end
```

jcechace commented on lib/rb_shift/service.rb:
```diff
@@ -7,19 +7,29 @@
+                        service: @name,
+                        **opts
+      end
+      if termination.nil?
```
Considering this the unless above is even worse.

```ruby
if termination
  # do the thing only when termination is NOT nil
else
  # do the thing only when termination IS nil
end
```

jcechace commented on lib/rb_shift/project.rb:
```diff
@@ -18,57 +18,86 @@ def self.open(name, client)
-      `oc new-app --#{source}=#{path} #{unfold_params(opts)} #{unfold_params(params, 'param')}`
-      invalidate
+    def new_app(source, path, block = false, timeout = 30, **opts)
+      execute 'new-app ', source.to_sym => path, **opts
```
Not sure if possible here, however prefer colon over rocket when constructing hashes

mijaros replied:
Well, you can't create symbol by colon from variable, for this you need to use the method
Source is variable which contains string name of source so there is to_sym method used properly.

jcechace commented on lib/rb_shift/client.rb:
```diff
@@ -1,8 +1,12 @@
 require_relative 'project'
 
 module RbShift
+  #
```
Nope

mijaros committed:
Fix missing comment in Client

Fix misused unless in service::create_route

Signed-off-by: Miroslav Jaros <mirek@mijaros.cz>

jcechace commented on lib/rb_shift/project.rb:
```diff
@@ -18,57 +18,86 @@ def self.open(name, client)
-      `oc new-app --#{source}=#{path} #{unfold_params(opts)} #{unfold_params(params, 'param')}`
-      invalidate
+    def new_app(source, path, block = false, timeout = 30, **opts)
+      execute 'new-app ', source.to_sym => path, **opts
```
U was talking about '=>' (rocket) vs ': (colon)' for hashes. The following should work just fine

```ruby
execute 'new-app ', source: path, **opts
```
Considering the following does wok

```ruby
a = 'foo'
{a: 'bar'}

# {:a=>"bar"}
```

jcechace commented on lib/rb_shift/client.rb:
```diff
@@ -81,5 +112,21 @@ def kube_entities
+      end.join(' ')
+    end
+
+    def unfold_params(params, prefix)
```
Is this method needed after ```unfold_opts``` was added?

jcechace commented on lib/rb_shift/client.rb:
```diff
@@ -22,30 +22,57 @@ def initialize(url, bearer_token)
-      `oc new-project #{name}`
+      execute "new-project #{name}", **opts
+      project = nil
+      project = projects(true).detect { |p| p.name == name } while project.nil?
```
Personally I would prefer ```find``` instead of ```detect``` but I might be convinced otherwise.

mijaros committed:
Remove unused unfold_params method

Replace detect with find

jcechace commented on lib/rb_shift/client.rb:
```diff
@@ -22,30 +22,56 @@ def initialize(url, bearer_token)
-      `oc new-project #{name}`
+      execute "new-project #{name}", **opts
+      project = nil
+      project = projects(true).find { |p| p.name == name } while project.nil?
```
Can't this possibly cause issue by repeatedly making api requests? ```load_projects``` doesn't have any time block.

mijaros replied:
I think that because reloading projects is quite time consuming operation, this cycle will in most cases take one iteration. Only when a cluster will be heavily used the operation will take two cycles.

jcechace commented on lib/rb_shift/client.rb:
```diff
@@ -81,5 +111,17 @@ def kube_entities
+        r = v.map { |m, n| "--#{k}=\"#{m}=#{n}\"" }.join(' ') if v.is_a? Hash
+        r = "--#{k}=#{v}" unless (v.is_a? Hash) || (v.is_a? Array)
+        r
+      end.join(' ')
```
Case statement would be more clean here, IMHO.

mijaros committed:
Change unfold_opts method to use case instead of conditionals

Modify client initialization to use either bearer_token or username,
password tuple
Fix some rubocop issues

Signed-off-by: Miroslav Jaros <mirek@mijaros.cz>

mijaros commented on lib/rb_shift/client.rb:
```diff
@@ -29,7 +29,6 @@ def initialize(url, bearer_token = nil, username = nil, password = nil)
                                              headers:    { Authorization: "Bearer #{bearer_token}" }
     end
 
-    # rubocop:disable Metrics/AbcSize
```
@jcechace This comment was marked by rubocop as unnecessary. This commit fixes it, but I'm not sure if this error does not appear on you version of rubocop. If this comment should stay, I will discard this commit, just let me know before merge.

mijaros commented on lib/rb_shift/client.rb:
```diff
@@ -29,7 +29,6 @@ def initialize(url, bearer_token = nil, username = nil, password = nil)
                                              headers:    { Authorization: "Bearer #{bearer_token}" }
     end
 
-    # rubocop:disable Metrics/AbcSize
```
@jcechace This comment was marked by rubocop as unnecessary. This commit fixes it, but I'm not sure if this error does not appear on you version of rubocop. If this comment should stay, I will discard this commit, just let me know before merge.

jcechace replied:
My version of rubocop unfortunately complains that the ABC complexity of ```get``` method is too high. I will fix the rubocop version in separate pull request.  Lets not mix things  together.

mijaros replied:
Ok, deleting this commit right away.

jcechace commented on lib/rb_shift/client.rb:
```diff
@@ -13,39 +13,71 @@ module RbShift
+      end
+
+      @token      = get_token(url, username, password) unless bearer_token
+      @token      = bearer_token if bearer_token
```
```ruby 
@token = bearer_token || get_token(url, username, password) 
```     
This would also be idiomatic python btw ;)

jcechace commented on lib/rb_shift/client.rb:
```diff
@@ -13,39 +13,71 @@ module RbShift
-      @token      = bearer_token
+    def initialize(url, bearer_token = nil, username = nil, password = nil)
+      if bearer_token.nil? && (username.nil? || password.nil?)
+        raise 'Token or username and password must be provided'
```
Custom exception would be more appropriate

jcechace commented:
Once the last batch of comments is addressed I think we can merge.  Assigning also to @pestanko to look over this.

mijaros committed:
Add custom ```InvalidAuthorizationError```

Fix token attribute setting in client

Signed-off-by: Miroslav Jaros <mirek@mijaros.cz>
